underscore.array.builders.js

```diff
@@ -191,6 +192,38 @@
       if (typeof obj == 'string')
         throw new TypeError('Strings cannot be reversed by _.reverseOrder');
       return slice.call(obj).reverse();
+    },
+    
+    
+    // Returns copy or array sorted according to arbitrary ordering
+    // order must be an array of values; defines the custom sort
+    // key must be one of: missing/null, a string, or a function; 
+    collate: function(array, order, key) {
+      if (typeof array.length != "number") throw new TypeError("expected an array-like first argument");
+      if (typeof order.length != "number") throw new TypeError("expected an array-like second argument");
+      
+      var original = slice.call(array);
+      var valA, valB;
+      return sort.call(original, function (a, b) {
+        if(_.isFunction(key)) {
+          valA = key.call(original, a);
+          valB = key.call(original, b);
+        } else if(existy(key)) {
+          valA = a[key];
+          valB = b[key];
+        } else {
+          valA = a;
+          valB = b;
+        }
+        
+        var rankA = _.indexOf(order, valA);
+        var rankB = _.indexOf(order, valB);
+        
+        if(rankA === -1) return 1;
+        if(rankB === -1) return -1;
```

Comment:
I agree with sorting unknown values after known values in principle. The way it is currently implemented, though, an unknown `a` will always be sorted after `b`, even if `b` is unknown as well. This makes the sorting more unstable than seems necessary to me. This can be addressed quite elegantly by reversing the order of these lines:

```suggestion
        if(rankB === -1) return -1;
        if(rankA === -1) return 1;
```

Reply:
I was wrong about this; your original code was already as stable as it could be, and my suggested change would break this. I have, however, added a test to check the stability.